lib/private/Repair/Apps.php

```
@@ -0,0 +1,221 @@
+ *
+ */
+
+namespace OC\Repair;
+
+use OC\RepairException;
+use OC_App;
+use OCP\App\AppAlreadyInstalledException;
+use OCP\App\AppManagerException;
+use OCP\App\AppNotFoundException;
+use OCP\App\AppNotInstalledException;
+use OCP\App\IAppManager;
+use OCP\Migration\IOutput;
+use OCP\Migration\IRepairStep;
+use OCP\Util;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use Symfony\Component\EventDispatcher\GenericEvent;
+
+class Apps implements IRepairStep {
+
+	const KEY_COMPATIBLE = 'compatible';
+	const KEY_INCOMPATIBLE = 'incompatible';
+	const KEY_MISSING = 'missing';
+
+	/** @var  IAppManager */
+	private $appManager;
+
+	/** @var  EventDispatcher */
+	private $eventDispatcher;
+
+	/**
+	 * Apps constructor.
+	 *
+	 * @param IAppManager $appManager
+	 */
+	public function __construct(IAppManager $appManager, EventDispatcher $eventDispatcher) {
+		$this->appManager = $appManager;
+		$this->eventDispatcher = $eventDispatcher;
+	}
+
+	/**
+	 * @return string
+	 */
+	public function getName() {
+		return 'Put back missing apps';
+	}
+
+	/**
+	 * @param IOutput $output
+	 * @throws RepairException
+	 */
+	public function run(IOutput $output) {
+		$appsToUpgrade = $this->getAppsToUpgrade();
+		$isMarketEnabled = $this->appManager->isEnabledForUser('market');
+		$failedCompatibleApps = [];
+		$failedIncompatibleApps = [];
+		$failedMissingApps = [];
+
+		if ($isMarketEnabled) {
+			$this->loadApp('market');
+			try {
+				$failedCompatibleApps = $this->getAppsFromMarket(
+					$output,
+					$appsToUpgrade[self::KEY_COMPATIBLE],
+					'upgradeAppStoreApp'
+				);
+				$failedIncompatibleApps = $this->getAppsFromMarket(
+					$output,
+					$appsToUpgrade[self::KEY_INCOMPATIBLE],
+					'upgradeAppStoreApp'
+				);
+				$failedMissingApps = $this->getAppsFromMarket(
+					$output,
+					$appsToUpgrade[self::KEY_MISSING],
+					'reinstallAppStoreApp'
+				);
+				$hasNotUpdatedCompatibleApps = count($failedCompatibleApps);
+
```

Hm, so the current implementation will try to enable apps that have been updated. https://github.com/owncloud/core/blob/2d000d9eba3815972241515d691ae7b8d49ce0c5/lib/private/App/AppManager.php#L218 currently might throw an \Exception, casing a full stop of this process, and then the rest of this function will not be executed.
1. I think it should throw an AppNotInstalledException instead.
2. That still allows the foreach to be escaped if the exception is thrown. I'd rather move 98-105 out of the try catch. If we try to enable an app that wasn't installed in the first place something has gone wrong and I think it is ok to fail early in that case.